src/AppInstallerCLICore/VTSupport.cpp

```diff
@@ -164,4 +164,63 @@ namespace AppInstaller::CLI::VirtualTerminal
         const Sequence EraseLineBackward{ AICLI_VT_CSI "1K" };
         const Sequence EraseLineEntirely{ AICLI_VT_CSI "2K" };
     }
+
+    namespace Progress
+    {
+        ConstructedSequence ProgressBar(ProgressState state, std::optional<int> percentage)
+        {
+            // See https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC
+
+            THROW_HR_IF(E_BOUNDS, percentage.has_value() && (percentage < 0 || percentage > 100));
+
+            // Workaround some quirks in the Windows Terminal implementation of the progress OSC sequence
+            switch (state)
+            {
+            case ProgressState::None:
+            case ProgressState::Indeterminate:
+                // Windows Terminal does not recognize the OSC sequence if the progress value is left out.
+                // As a workaround, we can specify an arbitrary value since it does not matter for None and Indeterminate states.
+                percentage = percentage.value_or(0);
+                break;
+            case ProgressState::Normal:
+            case ProgressState::Error:
+            case ProgressState::Paused:
+                // Windows Terminal does not support switching progress states without also setting a progress value at the same time,

```

Comment:
It seems like `Error` and `Paused` could also get the `value_or` treatment, since I don't think that there are any actual values that makes sense to pass in.

Reply:
Unless `Paused` is supposed to mean "paused at 50%", but I doubt that you are supposed to provide the HR to `Error`.

Reply:
> It seems like Error and Paused could also get the value_or treatment, since I don't think that there are any actual values that makes sense to pass in.

I think values do make sense for Error and Paused in general (not speaking about winget right now where we don't even use these states at the moment). E.g. your progress might be at 60% when an error occurs. In this case you may want to set the progress to 100% and state to Error at the same time.

Using Error or Paused with value zero does not look good because the user would basically see an empty progress bar colored in red or yellow. Paused is certainly meant to be used together with values (e.g. hit pause on a Windows file copy dialog), and Error would typically be set to 100 when it occurs to indicate the end of the failed operation. At least that's how I've seen it being used in some applications.

Reply:
I see, so the state and progress amount are somewhat independent.